feat(cli)!: find bots by default paths (#121)
ApeWorX/silverback: docs/userguides/development.md

You should have a python project with Silverback installed.
You can install Silverback via `pip install silverback`

## Project structure

The `silverback` cli automatically searches for python scripts to run as bots in specific locations relative to the root of your project.

> It if finds a module named `bot.py` either in the `bots/` directory or at the root directory of the project, then it will use that by default.

Reviewer: I dont think `bots/bot.py` should be recommended, I think it's either `bot.py`, `bot/__init__.py` (aka a module named `bot` for both cases), or a unique named script under `bots/`

Author: This requires a bit of changing for the dockerfile build

Author: Also, what if we allow them to name that folder whatever they want. If each bot requires a complicated project structure, shouldn't we allow them to use all of these features? Should we do this in a follow-up PR?

Reviewer: The only thing we need to do in the dockerfile is if they are using a script at path `bots/{name}.py` to move it to `bot.py` inside `Dockerfile.{name}` (and ignore the rest of the stuff in `bots/`)

Everything else is likely too complicated to be changing project structure around, and they should be making their own docker image manually

Author: I read through that paragraph and I'm not exactly suggesting to name your file `bot.py`. I'm stating that it automatically detects that file name. The rest of that section explains the different ways the project structure can be ran, do you want me to change that all together?

Reviewer: I want to remove `bots/bot.py` as autodetected

Author: Made the change, let me know if that works

You should have a python project with Silverback installed.
You can install Silverback via `pip install silverback`

## Project structure

There are 3 suggested ways to structure your project:

1. Create a `bot.py` file in the root directory of your project. This is the simplest way to define your bot project.

2. Create a `bots/` folder in the root directory of your project. Then develop bots in this folder as separate scripts (Do not include a __init__.py file).

> 3. Create a `bot/` folder with a `__init__.py` file that will include the instantiation of your `SilverbackApp()` object.

Reviewer 2: suggestion(non-blocking): There's a lot of repetition in this file. Especially around these 3 points. Consider consolidating a bit.

```

We will automatically detect all scripts under the `bots/` folder automatically, but if your bot resides in a location other than `bots/` then you can use this to run it:

```bash
silverback run folder.example:app --network your:network:of:choice
```

With a `bot/__init__.py` setup, you can run with:

```bash
silverback run bot --network your:network:of:choice
```

```{note}
It is suggested that you develop your modules as scripts to keep your deployments simple. If you have a deep understanding of containerization, and distributed applications, you can set your bots up however you'd like, and then create your own container definitions for deployments. For a simple deployment ecosystem, develop your bots as scripts, and avoid developing your applications as packages. (Do not include an __init__.py module, and do not use python modules for reusable code). If you follow this simple architecture, your deployments will be almost entirely hands off.
```

## Creating an Application

Creating a Silverback Application is easy, to do so initialize the `silverback.SilverbackApp` class:

```py
from silverback import SilverbackApp

> bot = SilverbackApp()

Reviewer: Question: what do we think about renaming `SilverbackApp` to `SilverbackBot`, to match the rest of this?

Author: seems like a major change, no? Like should bump the major for that, idk

Reviewer: This PR is already a minor bump, I was planning for v0.6.0

You should have a python project with Silverback installed.
You can install Silverback via `pip install silverback`

## Project structure

There are 3 suggested ways to structure your project:

1. Create a `bot.py` file in the root directory of your project. This is the simplest way to define your bot project.

2. Create a `bots/` folder in the root directory of your project. Then develop bots in this folder as separate scripts (Do not include a __init__.py file).

3. Create a `bot/` folder with a `__init__.py` file that will include the instantiation of your `SilverbackApp()` object.

The `silverback` cli automatically searches for python scripts to run as bots in specific locations relative to the root of your project.

> If `silverback` finds a module named `bot.py` either in the `bots/` directory or at the root directory of the project, then it will use that by default.

Reviewer: ```suggestion
If `silverback` finds a module named `bot` in the root directory of the project, then it will use that by default.
```

It will also be able to detect the scripts inside your `bots/` directory and let you run those by name (in case you have multiple bots in your project).

> `silverback` can also deal with a `bot/` folder with an `__init__.py` file that is defined as the runner.

Reviewer: Skip this as you mention below
```suggestion
```

```{note}
It is suggested that you create the instance of your `SilverbackApp()` object by naming the variable `bot`, since `silverback` will autodetect that variable name when loading your script file.
```

Another way you can structure your bot is to create a `bot` folder and define a runner inside of that folder as `__init__.py`.

If you have a more complicated project that requires multiple bots, naming each bot their own individual name is okay to do, and we encourage you to locate them under the `bots/` folder relative to the root of your project.

> This will work fairly seamless with the rest of the examples shown in this guide.

Reviewer: ```suggestion
This will work fairly seamlessly with the rest of the examples shown in this guide.
```

> To run a bot, as long as your project directory follows the suggestion above by using a `bot.py` script, you can run it easily with:

Reviewer: ```suggestion
To run a bot, as long as your project directory follows the suggestions above by using a `bot` module, you can run it easily with:
```

```bash
silverback run --network your:network:of:choice
```

However, if your bot's module name is not `bot.py` but `example.py` (for example), you can run it like this:

```bash
silverback run example --network your:network:of:choice
```

If the variable that you call the `SilverbackApp()` object is something other than `bot`, you can specific that by adding `:{variable-name}`:

```bash

> silverback run example:{variable-name} --network your:network:of:choice

Reviewer: ```suggestion
silverback run example:my_bot --network your:network:of:choice
```

```

We will automatically detect all scripts under the `bots/` folder automatically, but if your bot resides in a location other than `bots/` then you can use this to run it:

```bash
silverback run folder.example:app --network your:network:of:choice
```

> With a `bot/__init__.py` setup, you can run with:

Reviewer: ```suggestion
Note that with a `bot/__init__.py` setup, silverback will also autodetect it, and you can run it with:
```

```bash

> silverback run bot --network your:network:of:choice

Reviewer: ```suggestion
silverback run --network your:network:of:choice
```

```

```{note}

> It is suggested that you develop your modules as scripts to keep your deployments simple. If you have a deep understanding of containerization, and distributed applications, you can set your bots up however you'd like, and then create your own container definitions for deployments. For a simple deployment ecosystem, develop your bots as scripts, and avoid developing your applications as packages. (Do not include an __init__.py module, and do not use python modules for reusable code). If you follow this simple architecture, your deployments will be almost entirely hands off.

Reviewer: ```suggestion
It is suggested that you develop your bots as scripts to keep your deployments simple.
If you have a deep understanding of containerization, and have specific needs, you can set your bots up however you'd like, and then create your own container definitions for deployments to publish to your reqistry of choice.
For the most streamlined experience, develop your bots as scripts, and avoid relying on local packages
(e.g. do not include an `__init__.py` file inside your `bots/` directory, and do not use local modules inside `bots/` for reusable code).
If you follow these suggestions, your Silverback deployments will be easy to use and require almost no thought.
```

It will also be able to detect the scripts inside your `bots/` directory and let you run those by name (in case you have multiple bots in your project).
It is suggested that you create the instance of your `SilverbackApp()` object by naming the variable `bot`, since `silverback` will autodetect that variable name when loading your script file.

> Another way you can structure your bot is to create a `bot` folder and define a runner inside of that folder as `__init__.py`. This method of setting up your project will not allow you to use `silverback build` currently, and you will have to define your own dockerfile. There is a below example of how to run an app defined in this manner.

Reviewer: I think we can support this with the docker file though, just see if it's a `bot.py` or `bot` folder and copy over appropiately

the third case is that a script is under `bots/{name}.py`, which will get copied in the dockerfile from that location to `bot.py` (the first case which we support automatic loading for)

Reviewer: The endgame here is that each of the 3 "suggested" ways of defining a bot ends up with a local `bot` module inside of the dockerfile

Author: I'll organize this as a list in here
